tmexp/create_bow.py

```diff
@@ -0,0 +1,209 @@
+            if not (cur_blob is None and topic_model == HALL_MODEL):
+                prev_id = len(file_bows) - 1
+                if prev_blob == cur_blob:
+                    file_bows[prev_id]["refs"].append(ref)
+                else:
+                    file_bows[prev_id + 1] = {"refs": [ref]}
+                    if topic_model == HALL_MODEL:
+                        file_bows[prev_id + 1]["all"] = cur_word_counts
+                    elif topic_model == DIFF_MODEL:
+                        cur_word_counts.subtract(prev_word_counts)
+                        file_bows[prev_id + 1]["added"] = {
+                            word: count
+                            for word, count in cur_word_counts.items()
+                            if count > 0
+                        }
+                        file_bows[prev_id + 1]["removed"] = {
+                            word: -count
+                            for word, count in cur_word_counts.items()
+                            if count < 0
+                        }
+                        if cur_blob is None:
+                            prev_word_counts = Counter()
+                        else:
+                            prev_word_counts = word_counts[cur_blob]
+                    num_docwords += len(+cur_word_counts) + len(-cur_word_counts)
+                    prev_blob = cur_blob
+            prev_ref = ref
+        bow[file_path] = file_bows
+
+    if min_word_frac > 0 or max_word_frac < 1:
+        logger.info("Used %d blobs to create documents." % num_blobs)
+        min_word_blob = int(min_word_frac * num_blobs)
+        max_word_blob = int(max_word_frac * num_blobs)
+        logger.info(
+            "Finding words that appear in less then %d blobs or more then %d blobs ...",
+            min_word_blob,
+            max_word_blob,
+        )
+        blacklisted_words = {
+            word
+            for word, count in doc_freq.items()
+            if count < min_word_blob or count > max_word_blob
+        }
+        logger.info("Found %d words. Removing them ..." % len(blacklisted_words))
+        for file_path, file_bows in bow.items():
+            for index, ind_dict in file_bows.items():
+                for suffix, word_dict in ind_dict.items():
+                    if suffix == "refs":
+                        continue
+                    bow[file_path][index][suffix] = {
+                        word: count
+                        for word, count in word_dict.items()
+                        if word not in blacklisted_words
+                    }
+
+    logger.info("Creating word index ...")
+    sorted_vocabulary = sorted(doc_freq)
+    word_index = {word: i for i, word in enumerate(sorted_vocabulary)}
+    num_words = len(word_index)
+    logger.info("Number of distinct words: %d" % num_words)
+    logger.info("Saving word index ...")
+    with open(words_output_path, "w") as fout:
+        for word in sorted_vocabulary:
+            fout.write("%s \n" % word)
```

Comment:
there is an extra space I think. Also, we can write `fout.write("%s\n" % "\n".join(sorted_vocabulary))`.

Reply:
done